_posts/processes-state-evolution.md

```diff
@@ -0,0 +1,87 @@
+   Backend ->> Long-running process: Trigger long-running process
+   Long-running process ->> Database: Manipulate data
+   Long-running process ->> Backend: Process completed
+   Backend ->> Rest API: Topic event
+   Rest API ->> Web client: WebSocket message
+</div>
+
+In the [previous article](https://milestone.topics.it/2022/07/04/messages-evolution.html), we imagined that the new business requirement is to add an option for users to choose from multiple payment methods. We analyzed how the need affects message contracts and how to address the many challenges.
+
+For the sake of the discussion, let's now imagine a new business requirement: we must comply with GDPR requirements. To do so, we need to show users the privacy policy and keep track of whether users accepted the privacy policy for each OCR processing request.
+
+Sounds trivial. Ultimately, it's a boolean value that only needs a checkbox on the user interface. First, we need to carry that value around, and to do so, we need to update message contracts.
+
+Second, Let's look at how that affects a relational database.
+
+We could add a new column to the schema if we need to keep track of the privacy acceptance. A bit type sounds like it's enough. However, that comes with some headaches. If we change the schema, we immediately break the compatibility with running and completed processes. They know nothing about privacy acceptance. We assume they didn't accept the policy, or we decide to add the new column as nullable. If the value is null, the process is from when that requirement was non-existent; otherwise, we respect the set value.
+
+It's not pleasant, but it works. Let's now extend the requirement a bit. Imagine some countries do not require any privacy policy acceptance. That complicates the data schema a lot. Besides the nullable bit column, we need a country column. The new country column must be nullable, too, to be backward compatible.
+
+I guess you see the problem. As we implement new requirements, the code becomes an explosion of permutations to handle all the possible scenarios. A mitigation strategy could be having a column representing the process version. The code can make assumptions about the database schema based on the version number.
+
+> To be precise, the problem doesn't arise because of the rigid nature of relational databases. The same situation applies to any storage type. The problem manifests in code if we use a key-value store or a document database.
+
+## Is there any reason to store everything in the same table?
+
+The quick answer is no. We tend to do that because we're used to it and because requirements usually dictate processes' evolution or extensions. Those words are tricky and lead us to think there is an excellent reason to extend the current process instead of creating a brand new one. The object-orientated programming mindset also kicks in, telling us that inheritance is an excellent technique to evolve something. Extension, again.
```

Comment:
```suggestion
The quick answer is no. We tend to do that because we're used to it and because requirements usually dictate processes' evolution or extensions. Those words are tricky and lead us to think there is an excellent reason to extend the current process instead of creating a brand new one. The object-oriented programming mindset also kicks in, telling us that inheritance is an excellent technique to evolve something. Extension, again.
```